crates/cairo-profiler/src/trace_reader/function_trace_builder.rs

```diff
@@ -0,0 +1,267 @@
+use crate::trace_reader::FunctionName;
+use anyhow::{Context, Result};
+use cairo_lang_sierra::extensions::core::{CoreConcreteLibfunc, CoreLibfunc, CoreType};
+use cairo_lang_sierra::program::{GenStatement, Program, ProgramArtifact, StatementIdx};
+use cairo_lang_sierra::program_registry::ProgramRegistry;
+use cairo_lang_sierra_to_casm::compiler::CairoProgramDebugInfo;
+use itertools::{chain, Itertools};
+use regex::Regex;
+use std::collections::HashMap;
+use std::ops::AddAssign;
+use trace_data::TraceEntry;
+
+const MAX_TRACE_DEPTH: u8 = 100;
+
+pub struct ProfilingInfo {
+    pub functions_stack_traces: Vec<FunctionStackTrace>,
```

Comment:
It would mix abstraction layers. Here we operate on functions only, then we append the remaining entrypoint trace to the beggining of each function trace